docs/rfcs/0041-backend-native-api.md

```diff
@@ -0,0 +1,158 @@
+- Proposal Name: `backend_native_api`
+- Start Date: 2022-02-18
+- RFC PR: [datafuselabs/opendal#41](https://github.com/datafuselabs/opendal/pull/41)
+- Tracking Issue: [datafuselabs/opendal#0000](https://github.com/datafuselabs/opendal/issues/0000)
+
+# Summary
+
+Refactor API in backend native way to archive the performance.
+
+# Motivation
+
+## Poor performance
+
+`opendal` is relatively slow on `fs`: [performance drop 3 times after bump up opendal](https://github.com/datafuselabs/databend/issues/4197)
+
+First, we will do at least three syscalls in every' read' operation: `open`, `seek`, `read`.
+
+```rust
+let mut f = fs::OpenOptions::new()
+    .read(true)
+    .open(&path)
+    .await
+    .map_err(|e| parse_io_error(&e, &path))?;
+
+if let Some(offset) = args.offset {
+    f.seek(SeekFrom::Start(offset))
+        .await
+        .map_err(|e| parse_io_error(&e, &path))?;
+};
+```
+
+To make everything worse, our `SeelableReader` is designed for object storage systems and works much slower on local fs.
+
+`SeelableReader` will try to prefetch data in memory and maintain an internal pointer to implement `AsyncSeek`. However, due to our poor (nearly no) optimization, `SeekableReader` is much slower than using a system call `seek`.
+
+## Ergonomic unfriendly
+
+`opendal` is not easy to use either.
+
+In our early adoption project `databend`, we can see a lot of code looks like:
+
+```rust
+let data_accessor = self.data_accessor.clone();
+let path = self.path.clone();
+let reader = SeekableReader::new(data_accessor, path.as_str(), stream_len);
+let reader = BufReader::with_capacity(read_buffer_size as usize, reader);
+Self::read_column(reader, &col_meta, data_type.clone(), arrow_type.clone()).await
+```
+
+And
+
+```rust
+op.stat(&path).run().await
+```
+
+## Conclusion
+
+So in this proposal, I expect to address those problems. After implementing this proposal, we have a faster and easier-to-use `opendal`.
+
+# Guide-level explanation
+
+To operate on an object, we will use `Operator::object()` to create a new handler:
+
+```rust
+let o = op.object("path/to/file");
+```
+
+All operations that are available for `Object` for now includes:
```

Comment:
stat -> metadata
new_reader -> reader
new_writer -> writer

Since we have the corresponding `delete`, can we consider using `create` instead of `object`

Reply:
> stat -> metadata

LGTM, does `metadata` looks more native in rust?

> new_reader -> reader
> new_writer -> writer

I prefix a `new_` to indicate that we are creating a new struct `Reader`/`Writer`. Does it also obvious by using `o.reader()` and `o.writer()`?

> Since we have the corresponding `delete`, can we consider using `create` instead of `object`. Or, to avoid ambiguity with `new`, consider using `get`?

I'm afraid neither using `create` nor `get` is not good enough.

For `create`, its behavior is not so apparent. We have to explain that `create` doesn't mean creating an object on the underlying backend. We are just creating an object handle. 

For `get`, `op.get("path/to/object")` returns an `Object` to me is not so understandable. But if using `get_object`, it's easy to be confused with `get_object` operation in s3 alike services.

---

And in the future, I plan to add `list_objects` support via API `objects(path)`, so we will have:

```rust
let o = op.object("path/to/file")
// logic for object

let oi = op.objects("path/to/dir");
for o in io.iter() {
   // logic for object
}
```

Does it look nice for you?

Reply:
> LGTM, does `metadata` looks more native in rust?

We always say `file metadata`/`object metadata`, so I think it's more intuitive to name it that way. Rust should have no restrictions on naming.
 
> I prefix a `new_` to indicate that we are creating a new struct `Reader`/`Writer`. Does it also obvious by using `o.reader()` and `o.writer()`?

I think it's ok. this idea comes from the [bytes](https://docs.rs/bytes) crate .

https://docs.rs/bytes/latest/bytes/buf/trait.BufMut.html#method.writer

https://docs.rs/bytes/latest/bytes/buf/trait.Buf.html#method.reader

> And in the future, I plan to add `list_objects` support via API `objects(path)`, so we will have:

I see, so let's continue using `object` & `objects`. Or we can use `file` & `dir`.

Reply:
Updated in the following commits.